pg-extend/src/lib.rs

```diff
@@ -140,3 +151,24 @@ macro_rules! pg_create_stmt_bin {
         }
     };
 }
+
+/// Provides a barrier between Rust and Postgres' usage of the C set/longjmp
+///
+/// See the man pages for info on setjmp http://man7.org/linux/man-pages/man3/setjmp.3.html
+pub(crate) unsafe fn guard_c<R, F: FnOnce() -> R>(f: F) -> R {
```

Comment:
Per https://github.com/rust-lang/rfcs/issues/2625#issuecomment-460141114, I'm pretty sure any function that calls `setjmp` needs a `#[inline(never)]` directive, to avoid possible compiler mis-optimizations due to the fact that Rust doesn't yet tell LLVM that `setjmp` can return more than once.

Reply:
Currently I'm more worried about optimizations around `longjmp` than `setjmp` at least in terms of ordering. I think it's safe to throw `inline(never)` on this definitely. In the case of longjmp I'm thinking we're also going to need some compiler guards inserted as well: https://doc.rust-lang.org/std/sync/atomic/fn.compiler_fence.html

btw, if it wasn't clear, this review is still in a very early state. I just wanted to get what I was thinking down for conversation.